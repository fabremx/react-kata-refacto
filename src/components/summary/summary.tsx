import { ReactElement } from "react";
import { Product } from "../../models";
import styles from "./summary.module.scss";

interface Props {
  products: Product[];
  isUserVIP: boolean;
  coupon: number;
}

export function Summary({ products, isUserVIP, coupon }: Props): ReactElement {
  const calculTotal = (products: Product[]): number => {
    return products.reduce(
      (acc: number, product: Product): number => acc + product.price,
      0
    );
  };

  const calculTotalWithCouponDiscount = (total: number, coupon: number) => {
    if (!coupon) return total;
    return total - total * (coupon / 100);
  };

  const calculWithDeliveryFees = (total: number, isUserVIP: boolean) => {
    const DELIVERY_FEES = 3.99;

    if (!isUserVIP) return total + DELIVERY_FEES;
    return total;
  };

  const calculTotalAfterDiscount = (
    products: Product[],
    isUserVIP: boolean,
    coupon: number
  ): number => {
    const total: number = calculTotal(products);
    const totalAfterCoupon: number = calculTotalWithCouponDiscount(
      total,
      coupon
    );
    const totalAfterCouponAndDeliveryFees: number = calculWithDeliveryFees(
      totalAfterCoupon,
      isUserVIP
    );
    return totalAfterCouponAndDeliveryFees;
  };

  return (
    <div className={styles.summary}>
      <h3>Order summary</h3>

      <div className={styles.element}>
        <h4>Sub Total</h4>
        <div data-testid="subTotal">$ {calculTotal(products).toFixed(2)} </div>
      </div>

      <>
        <hr />
        {coupon && (
          <div className={styles.element}>
            <h4>Coupon Discount</h4>
            <div>% {coupon}</div>
          </div>
        )}

        <div className={styles.element}>
          <h4>Shipping Cost</h4>
          <div>
            {isUserVIP && <span>Free</span>}
            {!isUserVIP && <span>$ 3.99</span>}
          </div>
        </div>
      </>

      <div className={`${styles.element} ${styles.total}`}>
        <h5>Total</h5>
        <div data-testid="total">
          $ {calculTotalAfterDiscount(products, isUserVIP, coupon).toFixed(2)}
        </div>
      </div>
    </div>
  );
}
