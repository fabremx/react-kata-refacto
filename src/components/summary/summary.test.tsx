import { render } from "@testing-library/react";
import {
  COMMON_PRODUCT,
  ENDED_PRODUCT,
  NEW_PRODUCT_WITH_DISCOUNT,
} from "../../mocks/products";
import { Summary } from "./summary";

const PRODUCTS = [NEW_PRODUCT_WITH_DISCOUNT, ENDED_PRODUCT, COMMON_PRODUCT];
describe("Summary", () => {
  it("should render correctly the component when products is empty", () => {
    // When
    const { asFragment } = render(
      <Summary products={[]} isUserVIP={true} coupon={0} />
    );
    // Then
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render correctly the component when user is VIP", () => {
    // When
    const { asFragment } = render(
      <Summary products={PRODUCTS} isUserVIP={true} coupon={0} />
    );
    // Then
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render correctly the component when user is NOT VIP", () => {
    // When
    const { asFragment } = render(
      <Summary products={PRODUCTS} isUserVIP={false} coupon={0} />
    );
    // Then
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render correctly the component when user has coupon", () => {
    // When
    const { asFragment } = render(
      <Summary products={PRODUCTS} isUserVIP={false} coupon={10} />
    );
    // Then
    expect(asFragment()).toMatchSnapshot();
  });
});
