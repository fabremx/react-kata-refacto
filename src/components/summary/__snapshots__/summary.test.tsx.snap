// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary should render correctly the component when products is empty 1`] = `
<DocumentFragment>
  <div
    class="summary"
  >
    <h3>
      Order summary
    </h3>
    <div
      class="element"
    >
      <h4>
        Sub Total
      </h4>
      <div>
        $ 0.00 
      </div>
    </div>
    <hr />
    <div
      class="element total"
    >
      <h5>
        Total
      </h5>
      <div>
        $ 0.00
      </div>
    </div>
    <hr />
    <div
      class="submit"
    />
  </div>
</DocumentFragment>
`;

exports[`Summary should render correctly the component when user has coupon 1`] = `
<DocumentFragment>
  <div
    class="summary"
  >
    <h3>
      Order summary
    </h3>
    <div
      class="element"
    >
      <h4>
        Sub Total
      </h4>
      <div>
        $ 25.00 
      </div>
    </div>
    <div
      data-testid="discountDetails"
    >
      <hr />
      <div
        class="element"
      >
        <h4>
          Coupon Discount
        </h4>
        <div>
          % 10
        </div>
      </div>
      <div
        class="element"
      >
        <h4>
          Shipping Cost
        </h4>
        <div>
          <span>
            $ 3.99
          </span>
        </div>
      </div>
    </div>
    <hr />
    <div
      class="element total"
    >
      <h5>
        Total
      </h5>
      <div>
        $ 26.49
      </div>
    </div>
    <hr />
    <div
      class="submit"
    >
      <button
        data-testid="buyBtn"
      >
        Buy
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Summary should render correctly the component when user is NOT VIP 1`] = `
<DocumentFragment>
  <div
    class="summary"
  >
    <h3>
      Order summary
    </h3>
    <div
      class="element"
    >
      <h4>
        Sub Total
      </h4>
      <div>
        $ 25.00 
      </div>
    </div>
    <div
      data-testid="discountDetails"
    >
      <hr />
      0
      <div
        class="element"
      >
        <h4>
          Shipping Cost
        </h4>
        <div>
          <span>
            $ 3.99
          </span>
        </div>
      </div>
    </div>
    <hr />
    <div
      class="element total"
    >
      <h5>
        Total
      </h5>
      <div>
        $ 28.99
      </div>
    </div>
    <hr />
    <div
      class="submit"
    >
      <button
        data-testid="buyBtn"
      >
        Buy
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Summary should render correctly the component when user is VIP 1`] = `
<DocumentFragment>
  <div
    class="summary"
  >
    <h3>
      Order summary
    </h3>
    <div
      class="element"
    >
      <h4>
        Sub Total
      </h4>
      <div>
        $ 25.00 
      </div>
    </div>
    <div
      data-testid="discountDetails"
    >
      <hr />
      0
      <div
        class="element"
      >
        <h4>
          Shipping Cost
        </h4>
        <div>
          <span>
            Free
          </span>
        </div>
      </div>
    </div>
    <hr />
    <div
      class="element total"
    >
      <h5>
        Total
      </h5>
      <div>
        $ 25.00
      </div>
    </div>
    <hr />
    <div
      class="submit"
    >
      <button
        data-testid="buyBtn"
      >
        Buy
      </button>
    </div>
  </div>
</DocumentFragment>
`;
